### DISCUSSION TITLE
Breakpoints not hit while debugging

### DISCUSSION THREAD

**Discussion by Ing-Dom** [earlephilhower/arduino-pico#457] [Q&A] [answered]
I tried to get debugging with picoprobe to work under Windows 10.

While the openocd included in this does not work atm, there is also a [precompiled one from digikey](https://www.digikey.de/de/maker/projects/raspberry-pi-pico-and-rp2040-cc-part-2-debugging-with-vs-code/470abc7efb07432b82c95f6f67f184c0).

Debugging_RP2040\openocd\tcl>..\src\openocd.exe -f interface/picoprobe.cfg -f target/rp2040.cfg

With the gdb from this core:

k>C:\Users\dosp\AppData\Local\Arduino15\packages\rp2040\tools\pqt-gcc\1.3.2-a-9d55fd1\bin\arm-none-eabi-gdb.exe rp2040_blink.ino.elf

> C:\Users\dosp\AppData\Local\Arduino15\packages\rp2040\tools\pqt-gcc\1.3.2-a-9d55fd1\bin\arm-none-eabi-gdb.exe:
> **warning: Couldn't determine a path for the index cache directory.**
> GNU gdb (GDB) 8.2.50.20190202-git
> Copyright (C) 2019 Free Software Foundation, Inc.
> License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
> This is free software: you are free to change and redistribute it.
> There is NO WARRANTY, to the extent permitted by law.
> Type "show copying" and "show warranty" for details.
> This GDB was configured as "--host=i686-w64-mingw32 --target=arm-none-eabi".
> Type "show configuration" for configuration details.
> For bug reporting instructions, please see:
> <http://www.gnu.org/software/gdb/bugs/>.
> Find the GDB manual and other documentation resources online at:
>     <http://www.gnu.org/software/gdb/documentation/>.
> 
> For help, type "help".
> Type "apropos word" to search for commands related to "word"...
> Reading symbols from rp2040_blink.ino.elf...
> (gdb) target remote localhost:3333
> Remote debugging using localhost:3333
**> warning: Can not parse XML target description; XML support was disabled at compile time
> warning: Can not parse XML memory map; XML support was disabled at compile time**
> main () at C:\Users\dosp\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.10.0\cores\rp2040\main.cpp:100
> 100             if (arduino::serialEventRun) {
> (gdb) load
> Loading section .boot2, size 0x100 lma 0x10000000
> Loading section .text, size 0xd8c0 lma 0x10000100
> Loading section .rodata, size 0xbec lma 0x1000d9c0
> Loading section .ARM.extab, size 0xbc lma 0x1000e5ac
> Loading section .ARM.exidx, size 0x208 lma 0x1000e668
> Loading section .data, size 0x19ac lma 0x1000e870
> Start address 0x100001e8, load size 66076
> Transfer rate: 14 KB/sec, 7341 bytes/write.
> (gdb) monitor reset init
> (gdb) b main
> Breakpoint 1 at 0x100004bc: file C:\Users\dosp\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\1.10.0\cores\rp2040\main.cpp, line 56.
> (gdb) continue
> Continuing

=> pi pico is blinking, but the breakpoint was not hit. I retried several times also with setup, didn't work.
stepping works.

I highlighted some warnings that may be relevant?!

what setting in the board configuration triggers -g flag ?

"configuration": "flash=2097152_0,freq=125,opt=Debug,rtti=Enabled,dbgport=Disabled,dbglvl=All,usbstack=picosdk",

I tried to use another gbd binary, content of the official samd core (?):
Arduino15\packages\arduino\tools\arm-none-eabi-gcc\7-2017q4

And, with that gdb, debugging works like a charm...

also, the warnings displayed from the gdb.exe delivered with this core:

**> warning: Can not parse XML target description; XML support was disabled at compile time
> warning: Can not parse XML memory map; XML support was disabled at compile time**

are not displayed when i use the other gdb binary.

I'm hardly know anything about gcc toolchain, but that seems to be some problem with the binaries of this core...

**Comment by earlephilhower** [comment, marked as answer]
Are you sure you've got the right EL:F loaded in GDB?  I just tried under Linux on a simple test and things break properly.  Alternatively, maybe the OpenOCD you're using has some issues?    I've never seen those `XML support was disabled...` errors you've got.

This is under Linux with the latest repo at git head (which is 1.10.0 with 2 minor, unrelated API changes)

# openocd
````
earle@amd:~/Arduino/hardware/pico/rp2040/system$ ./openocd/bin/openocd -f interface/picoprobe.cfg -f target/rp2040.cfg
Open On-Chip Debugger 0.10.0+dev-ge3428fadb-dirty (2022-02-02-22:16)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
Info : only one transport option; autoselect 'swd'
Warn : Transport "swd" was already selected
adapter speed: 5000 kHz

Info : Hardware thread awareness created
Info : Hardware thread awareness created
Info : RP2040 Flash Bank Command
Info : Listening on port 6666 for tcl connections
Info : Listening on port 4444 for telnet connections
Info : clock speed 5000 kHz
Info : SWD DPIDR 0x0bc12477
Info : SWD DLPIDR 0x00000001
Info : SWD DPIDR 0x0bc12477
Info : SWD DLPIDR 0x10000001
Info : rp2040.core0: hardware has 4 breakpoints, 2 watchpoints
Info : rp2040.core1: hardware has 4 breakpoints, 2 watchpoints
Info : starting gdb server for rp2040.core0 on 3333
Info : Listening on port 3333 for gdb connections

Info : accepting 'gdb' connection on tcp/3333
target halted due to debug-request, current mode: Thread 
xPSR: 0x01000000 pc: 0x00000178 msp: 0x20041f00
target halted due to debug-request, current mode: Thread 
xPSR: 0x21000000 pc: 0x10001d62 msp: 0x20041fb0
Info : RP2040 B0 Flash Probe: 2097152 bytes @10000000, in 512 sectors

Info : New GDB Connection: 1, Target rp2040.core0, state: halted
target halted due to debug-request, current mode: Thread 
xPSR: 0xf1000000 pc: 0x000000ee msp: 0x20041f00
target halted due to debug-request, current mode: Thread 
xPSR: 0xf1000000 pc: 0x000000ee msp: 0x20041f00
target halted due to debug-request, current mode: Thread 
xPSR: 0x01000000 pc: 0x00000178 msp: 0x20041f00
````
# GDB
````
earle@amd:~/Arduino/hardware/pico/rp2040/system$ ./arm-none-eabi/bin/arm-none-eabi-gdb
GNU gdb (GDB) 8.2.50.20190202-git
Copyright (C) 2019 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "--host=x86_64-linux-gnu --target=arm-none-eabi".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word".
(gdb) target extended-remote :3333
Remote debugging using :3333
warning: No executable has been specified and target does not support
determining executable automatically.  Try using the "file" command.
0x10001d62 in ?? ()
(gdb) 
(gdb) file /tmp/arduino_
arduino_build_790722/ arduino_cache_777754/ 
(gdb) file /tmp/arduino_build_790722/
boot2.o                core/                  libraries/             preproc/               sketch_feb02b.ino.elf  sketch_feb02b.ino.uf2  
build.options.json     includes.cache         memmap_default.ld      sketch/                sketch_feb02b.ino.map  
(gdb) file /tmp/arduino_build_790722/sketch_feb02b.ino.elf
A program is being debugged already.
Are you sure you want to change the file? (y or n) y
Reading symbols from /tmp/arduino_build_790722/sketch_feb02b.ino.elf...
(gdb) b main
Breakpoint 1 at 0x100007f8: file /home/earle/Arduino/hardware/pico/rp2040/cores/rp2040/main.cpp, line 61.
(gdb) b loop
Breakpoint 2 at 0x10000348: file /home/earle/Arduino/sketch_feb02b/sketch_feb02b.ino, line 6.
(gdb) run
The program being debugged has been started already.
Start it from the beginning? (y or n) y
Starting program: /tmp/arduino_build_790722/sketch_feb02b.ino.elf 
Note: automatically using hardware breakpoints for read-only addresses.
target halted due to debug-request, current mode: Thread 
xPSR: 0x01000000 pc: 0x00000178 msp: 0x20041f00
[New Thread 2]

Thread 1 hit Breakpoint 1, main () at /home/earle/Arduino/hardware/pico/rp2040/cores/rp2040/main.cpp:61
61	    mutex_init(&_pioMutex);
(gdb) 
````

**Reply by Ing-Dom** [reply]
yes im pretty sure.
I also set up cortex-debug now (extension in visual studio code) and i got it working perfectly, but not with the binaries from your core, unfortunately.

And, I did the double check and configured cortex-debug (which uses openocd and gdb) to use "your" gdb.exe - and again, the same problem (warnings, no hitting breakpoints). changing the path to gdb back to the one shipped with arduino tools, and everyting is good again.

I assume it could be a problem with the crosscompiling?

Here you can see that the gdb from your core was built without Expat
I'm not sure if that causes the problem with the breakpoints, but I have no other idea...


> C:\Users\dosp>C:\Users\dosp\AppData\Local\Arduino15\packages\rp2040\tools\pqt-gcc\1.3.2-a-9d55fd1\bin\arm-none-eabi-gdb.exe
> C:\Users\dosp\AppData\Local\Arduino15\packages\rp2040\tools\pqt-gcc\1.3.2-a-9d55fd1\bin\arm-none-eabi-gdb.exe: warning: Couldn't determine a path for the index cache directory.
> GNU gdb (GDB) 8.2.50.20190202-git
> Copyright (C) 2019 Free Software Foundation, Inc.
> License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
> This is free software: you are free to change and redistribute it.
> There is NO WARRANTY, to the extent permitted by law.
> Type "show copying" and "show warranty" for details.
> This GDB was configured as "--host=i686-w64-mingw32 --target=arm-none-eabi".
> Type "show configuration" for configuration details.
> For bug reporting instructions, please see:
> <http://www.gnu.org/software/gdb/bugs/>.
> Find the GDB manual and other documentation resources online at:
>     <http://www.gnu.org/software/gdb/documentation/>.
> 
> For help, type "help".
> Type "apropos word" to search for commands related to "word".
> (gdb) show configuration
> This GDB was configured as follows:
>    configure --host=i686-w64-mingw32 --target=arm-none-eabi
>              --with-auto-load-dir=$debugdir:$datadir/auto-load
>              --with-auto-load-safe-path=$debugdir:$datadir/auto-load
>              **--without-expat**
>              --with-gdb-datadir=/workdir/arm-none-eabi.win32/share/gdb (relocatable)
>              --with-jit-reader-dir=/workdir/arm-none-eabi.win32/lib/gdb (relocatable)
>              --without-libunwind-ia64
>              --without-lzma
>              --without-babeltrace
>              --without-intel-pt
>              --disable-libmcheck
>              --without-mpfr
>              --without-python
>              --without-guile
>              --with-separate-debug-dir=/workdir/arm-none-eabi.win32/lib/debug (relocatable)
> 
> ("Relocatable" means the directory can be moved with the GDB installation
> tree, and GDB will still find it.)
> (gdb)

**Reply by earlephilhower** [reply]
I'm very doubtful it has anything to do with cross compiling.  The builds are passed the exact same params, just use different gcc versions to build cross.  It's more likely you're running something that doesn't mesh w/the GDB or picoprobe versions used.  I have never used cortex-debug, so can't comment on it's compatibility.

**Reply by Ing-Dom** [reply]
did you build it with expat ?
what does your linux gdb version print at "show configuration ?
you don't think the missing expat is the problem?

**If** it is not Expat, but some compatibility issue, then it could be the picoprobe version or the openocd version I use.
I use that picoprobe.uf2: https://www.raspberrypi.com/documentation/microcontrollers/raspberry-pi-pico.html#debugging-using-another-raspberry-pi-pico
and openocd from here. https://www.dropbox.com/sh/xz9s3qxxelstd66/AABy2xlym93c_E__7rYeqtwaa?dl=0

I can't be cortex-debug, because I have this issue also when I debug with plain gdb in the console...

**Reply by earlephilhower** [reply]
You need a custom build of openocd to work properly w/picoprobe.  For some reason the RPi guys forked and made a custom branch (rp2040) but never submitted a PR upstream to the main openocd.  That could be the issue, but if I were you I wouldn't waste time at this point.

You could try building the RPi fork of OpenOCD on your own box, if you want to give debugging a go: https://github.com/raspberrypi/openocd

**Reply by Ing-Dom** [reply]
this IS a custom build of openocd for the picoprobe (by digikey) I use.
otherwise it wouldn't work with the arm-none-eabi-gdb.exe provided by the Arduino SAMD boards package wouldn't it ?

Did you have a look about Expat in your working linux version of gdb ?

**Reply by earlephilhower** [reply]
Sorry, but I don't really know what Digikey has built.  The GDB that's shipped does have expat included I believe.  I'm not sure it makes any difference, but it threw me off seeing it in your log.

Again, the included Linux build works fine and the Windows build has the esxact same build `./configure` options but (as you found) the libusb update stuff breaks things sue to the changed gcc.dll.  I'm messing around w/a new Win10 VM to manually try building things but no promises yet!

Net net, if GDB is setting a break on the proper address, and you restart and don't see it break at that address, then there's something up w/the openocd<->picoprobe<->UUT configuration.  GDB's commands to a remote target are standardized, it's the OpenOCD that converts the generic "break here" to bits on a wire somewhere.

**Reply by Ing-Dom** [reply]
I'm really curious now what is the problem..

thats what i found out (as mentioned, just making 100% sure we don't have a misunderstanding)

gdb (from your core) <-> openocd (digikey) <-> picoprobe <-> pi pico  ==> fails
gdb (from arduino) <-> openocd (digikey) <-> picoprobe  <-> pi pico==> works

disreagrding GDB, it's 100% the same setup.

from my point of view (the dumb user), it must be the gdb what causes the problem, OR some incompatibility to openocd.
But - you said it yourself - the communication between gdb and openocd ist standardized, so I would not assume that.

But maybe there is some magic happening here thats beyond my (limited) imagination.

**Reply by earlephilhower** [reply]
````
earle@amd:~/src/pico-quick-toolchain$ ~/Arduino/hardware/pico/rp2040/system/arm-none-eabi/bin/arm-none-eabi-gdb 
GNU gdb (GDB) 8.2.50.20190202-git
Copyright (C) 2019 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "--host=x86_64-linux-gnu --target=arm-none-eabi".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word".
(gdb) show configuration
This GDB was configured as follows:
   configure --host=x86_64-linux-gnu --target=arm-none-eabi
             --with-auto-load-dir=$debugdir:$datadir/auto-load
             --with-auto-load-safe-path=$debugdir:$datadir/auto-load
             --with-expat
             --with-gdb-datadir=/workdir/arm-none-eabi.x86_64/share/gdb (relocatable)
             --with-jit-reader-dir=/workdir/arm-none-eabi.x86_64/lib/gdb (relocatable)
             --without-libunwind-ia64
             --without-lzma
             --without-babeltrace
             --without-intel-pt
             --disable-libmcheck
             --with-mpfr
             --without-python
             --without-guile
             --with-separate-debug-dir=/workdir/arm-none-eabi.x86_64/lib/debug (relocatable)

("Relocatable" means the directory can be moved with the GDB installation
tree, and GDB will still find it.)
(gdb) 
````

GDB's "set a breakpoint" API is an insanely simple text string.  I had to implement it (well, mostly debug and integrate 2 pre-existing setups) on the ESP8266.  https://www.embecosm.com/appnotes/ean4/embecosm-howto-rsp-server-ean4-issue-2.html#id3080572)

I think you are on to something w/Expat, though.  The cross-compile environment may not have libexpat.so installed and https://sourceware.org/gdb/onlinedocs/gdb/Requirements.html has this tidbit...

    GDB can use the Expat XML parsing library. This library may be included with your operating system distribution; if it is not, you can get the latest version from http://expat.sourceforge.net/. The configure script will search for this library in several standard locations; if it is installed in an unusual path, you can use the --with-libexpat-prefix option to specify its location.

    Expat is used for:

        Remote protocol memory maps (see [Memory Map Format](https://sourceware.org/gdb/onlinedocs/gdb/Memory-Map-Format.html#Memory-Map-Format))
        Target descriptions (see [Target Descriptions](https://sourceware.org/gdb/onlinedocs/gdb/Target-Descriptions.html#Target-Descriptions))
        Remote shared library lists (See [Library List Format](https://sourceware.org/gdb/onlinedocs/gdb/Library-List-Format.html#Library-List-Format), or alternatively see [Library List Format for SVR4 Targets](https://sourceware.org/gdb/onlinedocs/gdb/Library-List-Format-for-SVR4-Targets.html#Library-List-Format-for-SVR4-Targets))
        MS-Windows shared libraries (see [Shared Libraries](https://sourceware.org/gdb/onlinedocs/gdb/Files.html#Shared-Libraries))
        Traceframe info (see [Traceframe Info Format](https://sourceware.org/gdb/onlinedocs/gdb/Traceframe-Info-Format.html#Traceframe-Info-Format))
        Branch trace (see [Branch Trace Format](https://sourceware.org/gdb/onlinedocs/gdb/Branch-Trace-Format.html#Branch-Trace-Format), see [Branch Trace Configuration Format](https://sourceware.org/gdb/onlinedocs/gdb/Branch-Trace-Configuration-Format.html#Branch-Trace-Configuration-Format))

**Reply by earlephilhower** [reply]
Fixed in 1.12.0

**Comment by maxgerhardt** [comment]
I don't know what fixed it, but in the very latest Toolchain + Core version, this actually **works** for me in PlatformIO. Integration update soon.

![grafik](https://user-images.githubusercontent.com/26485477/172935348-f7ad04c6-40f8-4f60-a314-512616bda962.png)

Exciting!!

**Reply by earlephilhower** [reply]
The original problem was lack of XML support in the GDB build.  That was fixed several months ago, though, with https://github.com/earlephilhower/pico-quick-toolchain/commit/ed6d983b583499215aaa9fe68ea18720fc6993e7.

**Reply by Ing-Dom** [reply]
> Integration update soon.
looking forward to it.

does it work with pico-probe ?

**Reply by maxgerhardt** [reply]
Yes, see https://github.com/platformio/platform-raspberrypi/pull/36

In fact the above scerenshot was made **using** a Picoprobe.
